src/ncp/ncp_uart.cpp

```diff
@@ -47,138 +47,163 @@ extern "C" void otNcpInit(void)
 {
-    Reset();
+    Clear();
 }
 
-void
-NcpUart::SendHdlcBuffer::Reset(void)
+void NcpUart::UartTxBuffer::Clear(void)
 {
     mWritePointer = mBuffer;
     mRemainingLength = sizeof(mBuffer);
 }
 
-uint16_t
-NcpUart::SendHdlcBuffer::GetLength(void) const
+bool NcpUart::UartTxBuffer::IsEmpty(void) const
 {
-    return static_cast<uint16_t>(mWritePointer - mBuffer);
+    return mWritePointer == mBuffer;
 }
 
-const uint8_t *
-NcpUart::SendHdlcBuffer::GetBuffer(void) const
+uint16_t NcpUart::UartTxBuffer::GetLength(void) const
 {
-    return mBuffer;
+    return static_cast<uint16_t>(mWritePointer - mBuffer);
 }
 
-uint16_t
-NcpUart::SendHdlcBuffer::GetRemainingLength(void) const
+const uint8_t *NcpUart::UartTxBuffer::GetBuffer(void) const
 {
-    return mRemainingLength;
+    return mBuffer;
 }
 
 NcpUart::NcpUart():
     NcpBase(),
-    mFrameDecoder(mReceiveFrame, sizeof(mReceiveFrame), &HandleFrame, this),
-    mSendFrame()
+    mFrameDecoder(mRxBuffer, sizeof(mRxBuffer), &HandleFrame, this),
+    mUartBuffer(),
+    mTxFrameBuffer(mTxBuffer, sizeof(mTxBuffer)),
+    mUartSendDoneTask(HandleUartSendDone, this)
 {
+    mState = kStartingFrame;
+
+    mTxFrameBuffer.SetCallbacks(NULL, TxFrameBufferHasData, this);
 }
 
-uint16_t
-NcpUart::OutboundFrameGetRemaining(void)
+ThreadError NcpUart::OutboundFrameBegin(void)
 {
-    return mSendFrame.GetRemainingLength();
+    return mTxFrameBuffer.InFrameBegin();
 }
 
-ThreadError
-NcpUart::OutboundFrameBegin(void)
+ThreadError NcpUart::OutboundFrameFeedData(const uint8_t *aDataBuffer, uint16_t aDataBufferLength)
 {
-    ThreadError errorCode;
-
-    mSendFrame.Reset();
+    return mTxFrameBuffer.InFrameFeedData(aDataBuffer, aDataBufferLength);
+}
 
-    errorCode = mFrameEncoder.Init(mSendFrame);
+ThreadError NcpUart::OutboundFrameFeedMessage(Message &aMessage)
+{
+    return mTxFrameBuffer.InFrameFeedMessage(aMessage);
+}
 
-    return errorCode;
+ThreadError NcpUart::OutboundFrameSend(void)
+{
+    return mTxFrameBuffer.InFrameEnd();
 }
 
-ThreadError
-NcpUart::OutboundFrameFeedData(const uint8_t *frame, uint16_t frameLength)
+void NcpUart::TxFrameBufferHasData(void *aContext, NcpFrameBuffer *aNcpFrameBuffer)
 {
-    ThreadError errorCode;
+    (void)aContext;
+    (void)aNcpFrameBuffer;
 
-    errorCode = mFrameEncoder.Encode(frame, frameLength, mSendFrame);
+    sNcpUart->TxFrameBufferHasData();
+}
 
-    return errorCode;
+void NcpUart::TxFrameBufferHasData(void)
+{
+    if (mUartBuffer.IsEmpty())
+    {
+        EncodeAndSendToUart();
+    }
 }
 
-ThreadError
-NcpUart::OutboundFrameFeedMessage(Message &message)
+// This method encodes a frame from the tx frame buffer (mTxFrameBuffer) into the uart buffer and sends it over uart.
+// If the uart buffer gets full, it sends the current encoded portion. This method remembers current state, so on
+// sub-sequent calls, it restarts encoding the bytes from where it left of in the frame .
+void NcpUart::EncodeAndSendToUart(void)
```

Comment:
(The diff here looks a bit weird, so it may be better to look at the final implementation directly)
The code in `EncodeAndSendToUart()` ended up looking very `protothread`-like  :)